module/os/linux/spl/spl-dcache.c

```diff
@@ -0,0 +1,48 @@
+/*
+ *  Copyright (C) 2019 OpenZFS Project
+ *  Written by Pavel Snajdr <snajpa@snajpa.net>
+ *
+ *  This file is part of the SPL, Solaris Porting Layer.
+ *  For details, see <http://zfsonlinux.org/>.
+ *
+ *  The SPL is free software; you can redistribute it and/or modify it
+ *  under the terms of the GNU General Public License as published by the
+ *  Free Software Foundation; either version 2 of the License, or (at your
+ *  option) any later version.
+ *
+ *  The SPL is distributed in the hope that it will be useful, but WITHOUT
+ *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ *  for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with the SPL.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <sys/debug.h>
+#include <sys/dcache.h>
+#include <sys/sysmacros.h>
+#include <sys/kmem.h>
+#include <sys/vmem.h>
+#include <linux/mm.h>
+#include <linux/dcache.h>
+#include <linux/list.h>
+
+/*
+ * Locklessly walk and invalidate all aliases of an inode; the lockless
+ * approach assumes the function is only called when z_teardown_lock is
+ * active or the inode in question is locked.
+ */
+void
+spl_d_drop_aliases(struct inode *inode)
```

Comment:
This is a good idea for how to get rid of the dentries, and I think you're on the right track.  Could you explain why it was necessary to do this locklessly?  Did you encounter a deadlock when using `d_drop()`?

Reply:
Yes, though I didn't spare much time investigating with what did it exactly deadlock; NFS code uses __d_drop in one instance too, when it needs to make a VFS entry disappear, so I just copied that approach.

Reply:
Actually, it turns out, that there's no deadlock, at least I hope so. I'll run my current version through the bots and we'll see.